### PR TITLE
feat(prototypes): mention scenario with dropdown

### PR DESCRIPTION
**Pull request by bmdalex**
### Description

This PR attempts to create a prototype to show users how they can mention people using `@` key in a rich text editor.

I am using a `contentEditable` div as rich text input and the `Dropdown` component to leverage the list of users and filtering through them. The rest of the functionality is in the prototype section 

### Screenshot
screen recording 2019-02-14 at 16 51 35

### Issues:
- #900 and issue in [description](https://github.com/stardust-ui/react/pull/900#issue-253086393) - **blocker**
- styling regression when using the filtering capability caused by #679 

### Later edit:
all issues and blocking items for this PR are fixed

### CODE REVIEW

File: docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx [hunk 1]
```diff
@@ -34,25 +34,55 @@ const createEntry = (): Entry => ({
...
   }
 
-  handleSearchQueryChange = (e: React.SyntheticEvent, { searchQuery }: DropdownProps) => {
+  private handleSearchQueryChange = (e: React.SyntheticEvent, { searchQuery }: DropdownProps) => {
     this.setState({ searchQuery })
     this.fetchItems()
   }
 
-  fetchItems = () => {
+  private fetchItems = () => {
```
Comment (bmdalex) [inline docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx]: https://github.com/stardust-ui/react/pull/901#discussion_r256900108 from @layershifter :

_What is purpose of these changes? Do we need them?
We don't have an agreement on visibility modifiers so let's revert these changes._

Comment (bmdalex) [reply docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx]: @layershifter 

> We don't have an agreement on visibility modifiers so let's revert these changes.

I don't get this part, it's pretty obvious to me that in Typescript class members that are not meant to be used outside the scope of the class are private.

> What is purpose of these changes? Do we need them?

I will revert the changes however because the prototype section does not have the JS version so you might argue the change does not bring any benefit here.

File: docs/src/prototypes/dropdowns/utils.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    return
+  }
+
+  const elem = document.createElement('span')
+  elem.id = id
+  if (text) {
+    elem.innerText = text
+  }
+  range.insertNode(elem)
+}
```
Comment (bmdalex) [inline docs/src/prototypes/dropdowns/utils.ts:51]: https://github.com/stardust-ui/react/pull/901#discussion_r256902962 from @layershifter 

Is there any third party dependency that does this work? That is fully tested and supported?
Why we can't use [`caret-pos`](https://www.npmjs.com/package/caret-pos)?

Comment (bmdalex) [reply docs/src/prototypes/dropdowns/utils.ts:51]: @layershifter Just wanted to keep things minimal and avoid  introducing extra dependencies just for a prototype, clients using this code will end up needing it too.

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      />
+    )
+  }
+
+  private showDropdown = () => {
+    this.dropdownExists = true
+    insertNodeAtCursorPosition({ id: this.mountNodeId })
+
+    const node = this.getMountNode()
+    ReactDOM.render(
```
Comment (bmdalex) [inline docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: https://github.com/stardust-ui/react/pull/901#discussion_r256901115 from @layershifter 

Can you please why we can't use `createPortal()`/`PortalInner` component there? What is a reason for another React's root?

Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: no reason, 1st iteration, let me try that, thanks for the suggestion

File: docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx (hunk 1 above)
Comment (kuzhelov) [reply docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx]: agree with both sides here. Actually, if we would be absolutely pedantic here and adhere to the flavour of OO paradigm introduced by TS, we indeed should use scope modifiers for components' methods - to showcase which ones meant to be accessible from outside as part of public API.

At the same time, the main objection around having this burden of scope modifiers was reasonable fact that this is pretty rare scenario (that, almost for sure would be invalid in Stardust's case) when we would need to access methods of instantiated React component from outside world. Given this reasonable fact, it was considered to be unnecessary to have this scope modifiers being provided to React components.

That being said, just for the sake of moving forward, lets stick to the thing we've agreed upon (i.e. not introduce scope modifiers for React components) - although, as @Bugaa92 has noticed, this is the prototype's code where this is reasonable approach. And, if there is still strong objection against that (or, for certain, if there is any use case that showcases importance of having these modifiers) - @Bugaa92, please, feel free to schedule some time to discuss that 👍

Comment (layershifter) [reply docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx]: Personally for me: there is no big difference, but I _expect_ to see handlers and helpers method before `render()` methods as it's done in React docs: https://reactjs.org/docs/handling-events.html

File: docs/src/prototypes/dropdowns/dataMocks.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+export interface AtMention {
+  header: string
+  image: string
+  content: string
+}
+
+export const atMentionItems: AtMention[] = [
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/dataMocks.ts]:
```suggestion
export const atMentionItems: AtMentionItem[] = [
```

Type name is not represent the object, `AtMentionItem` should be more clear

File: docs/src/prototypes/dropdowns/dataMocks.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  {
+    header: 'Peter Parker',
+    image: 'public/images/avatar/small/daniel.jpg',
+    content: 'Partner Software Engineer',
+  },
+  {
+    header: 'Selina Kyle',
+    image: 'public/images/avatar/small/ade.jpg',
+    content: 'Graphic Designer',
+  },
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/dataMocks.ts]: Actually, we can use `faker` to create these data. Can we use it?

File: docs/src/prototypes/dropdowns/index.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import * as React from 'react'
+import { PrototypeSection, ComponentPrototype } from '../Protoypes'
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/index.tsx]:
```suggestion
import { PrototypeSection, ComponentPrototype } from '../Prototypes'
```

Oops, a typo ⌨️

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+class InputWithDropdownExample extends React.Component<{}, InputWithDropdownState> {
+  private readonly mountNodeId = 'dropdown-mount-node'
+  private readonly dropdownInputSelector = `#${this.mountNodeId} .${Input.slotClassNames.input}`
+  private readonly editorStyle: React.CSSProperties = {
+    backgroundColor: 'lightgrey',
+    padding: '5px',
+    minHeight: '100px',
+    outline: 0,
+  }
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: `editorStyle` can be a usual variable

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    this.setState({ searchQuery, dropdownValue })
+  }
+
+  private handleSearchQueryChange = (e: React.SyntheticEvent, { searchQuery }: DropdownProps) => {
+    this.setInputElementSize(searchQuery.length)
+    this.setState({ searchQuery })
+  }
+
+  private handleInputKeyDown = (e: React.KeyboardEvent) => {
+    const code = keyboardKey.getCode(e)
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/inputWithDropdown.tsx]:
```suggestion
    const keyCode = keyboardKey.getCode(e)
```

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        noResultsMessage="We couldn't find any matches."
+      />,
+      node,
+    )
+  }
+
+  private hideDropdownAndRestoreEditor = () => {
+    removeElement(this.getMountNode())
+
+    this.tryFocusEditor()
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/inputWithDropdown.tsx]:
```ts
_.invoke(this.contendEditableRef.current, 'focus')
```

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+    return ReactDOM.createPortal(
+      <Dropdown
+        defaultOpen={true}
+        inline
+        search
+        items={atMentionItems}
+        toggleIndicator={null}
+        searchInput={{
+          input: { autoFocus: true },
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: Can we avoid `setInputElementSize()` and just pass `size` as prop?

```ts
searchInput={{
  input: {
    autoFocus: true,
    size: dropdownOpen ? 0 : searchQuery.length.
  },
}}

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx (hunk 3 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: implemented using `createPortal`, works as expected, good call

File: docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx (hunk 1 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx]: let's close for now and stick to what we have, tracking in #933

File: docs/src/prototypes/dropdowns/dataMocks.ts (hunk 4 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/dataMocks.ts]: done, thanks

File: docs/src/prototypes/dropdowns/dataMocks.ts (hunk 5 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/dataMocks.ts]: done, thanks

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx (hunk 7 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: done, thanks

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx (hunk 8 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: done, thanks

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx (hunk 9 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: done, thanks

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx (hunk 10 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: yup, with the portal approach it's possible, good call; when I was rendering another node it didn't work; also, size needs to be min 1 so I did `searchQuery.length + 1`

File: docs/src/prototypes/Prototypes.tsx [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+interface PrototypeSectionProps {
+  title?: string
+}
+
+interface ComponentPrototypeProps extends PrototypeSectionProps {
+  description?: string
+}
+
+export const PrototypeSection: React.SFC<ComponentPrototypeProps> = props => (
```
Comment (layershifter) [inline docs/src/prototypes/Prototypes.tsx]:
```suggestion
export const PrototypeSection: React.FC<ComponentPrototypeProps> = props => (
```

SFC was deprecated :)

File: docs/src/prototypes/Prototypes.tsx [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  <Box style={{ margin: 20 }}>
+    {props.title && <Header as="h1">{props.title}</Header>}
+    {props.children}
+  </Box>
+)
+
+export const ComponentPrototype: React.SFC<ComponentPrototypeProps> = props => (
```
Comment (layershifter) [inline docs/src/prototypes/Prototypes.tsx]:
```suggestion
export const ComponentPrototype: React.FC<ComponentPrototypeProps> = props => (
```

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  padding: '5px',
+  minHeight: '100px',
+  outline: 0,
+}
+
+class InputWithDropdown extends React.Component<{}, InputWithDropdownState> {
+  private readonly initialState: InputWithDropdownState = {
+    dropdownOpen: false,
+    searchQuery: '',
+  }
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: NIT: `initialState` can be extracted to a variable as it doesn't have any coupling with the class

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+  render() {
+    return (
+      <>
+        <div
+          contentEditable
+          ref={this.contendEditableRef}
+          onKeyUp={this.handleEditorKeyUp}
+          style={editorStyle}
+        />
+        {this.renderDropdown()}
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: NIT: Do we really need `renderDropdown()`, it looks quite simple to be inlined

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+          onSearchQueryChange={this.handleSearchQueryChange}
+          noResultsMessage="We couldn't find any matches."
+        />
+      </CustomPortal>
+    )
+  }
+
+  private hideDropdownAndRestoreEditor = (cb?: () => void) => {
+    this.setState(this.initialState, () => {
+      this.tryFocusEditor()
```
Comment (layershifter) [inline docs/src/prototypes/dropdowns/inputWithDropdown.tsx]:
```suggestion
      _.invoke(this.contendEditableRef.current, 'focus')
```

NIT: `tryFocusEditor()`  used only once, may be we can remove it?

File: packages/react/src/components/Dropdown/DropdownSearchInput.tsx [hunk 16]
```diff
@@ -122,8 +124,8 @@ class DropdownSearchInput extends UIComponent<ReactProps<DropdownSearchInputProp
           onBlur: this.handleInputBlur,
           onKeyDown: this.handleInputKeyDown,
           ...accessibilityInputProps,
+          ...unhandledProps.input,
```
Comment (layershifter) [inline packages/react/src/components/Dropdown/DropdownSearchInput.tsx:127]: Looks weird, but I understand _why_ it's done.

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx (hunk 15 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: I think this is a matter of personal preference; I prefer keeping it more verbose, it makes it obvious to the reader what the code does without to understand details.

File: packages/react/src/components/Dropdown/DropdownSearchInput.tsx (hunk 16 above)
Comment (bmdalex) [reply packages/react/src/components/Dropdown/DropdownSearchInput.tsx:127]: yea..no better idea at this point

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx (hunk 14 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: with the latest refactoring..no, we don't need it, inlined it, thanks

File: docs/src/prototypes/Prototypes.tsx (hunk 12 above)
Comment (bmdalex) [reply docs/src/prototypes/Prototypes.tsx]: done, thanks

File: docs/src/prototypes/Prototypes.tsx (hunk 11 above)
Comment (bmdalex) [reply docs/src/prototypes/Prototypes.tsx]: done, thanks

File: docs/src/prototypes/dropdowns/inputWithDropdown.tsx (hunk 13 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/inputWithDropdown.tsx]: agreed on the fact there is no coupling, but the concept of `state` doesn't make sense outside of the component...

with that argument (`it doesn't have any coupling with the class`) most of `readonly` members in TS do not have any coupling with their class (are merely constants) but we add them as members because they make sense just in that class

File: docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx [hunk 17]
```diff
@@ -68,33 +68,26 @@ class AsyncDropdownSearch extends React.Component<{}, SearchPageState> {
...
-            searchQuery={searchQuery}
-            toggleIndicator={false}
-            value={value}
-          />
-          <Divider />
-          <CodeSnippet mode="json" value={this.state} />
-        </Segment>
-      </div>
+      <>
+        <Dropdown
```
Comment (kuzhelov) [inline docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx]: for this experience might suggest to introduce
- comment to example's code
- maybe better additional `<p>` in between of `Dropdown` and `CodeSnippet`, or add text to `Divider`

that would explain that `CodeSnippet` part rendered is necessary to reflect state of `fetchedData` - otherwise it was a bit confusing to me an, frankly, originally I thought that this is just a docs bug :) In either case, it would be helpful to explicitly say to the client that this is aimed to represent state of _fetched_ data.

image

Another small suggestion might be about reducing amount of data entries fetched - I think that even 5 (with names that would suggest reasonably good distribution) should work fine :) Key benefit is that it won't be necessary for the client to make long scroll to get to the next example (if this data section would be collapsable, it would help as well).

Comment (kuzhelov) [reply docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx]: also, when there is nothing found in the list, there is the following view of Dropdown produced:

image

This representation (with narrow white ribbon below the search field) seems to be a bit confusing - I think that we could easily fix it with `noResultsMessage` prop:

image

File: docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+const editorStyle: React.CSSProperties = {
+  backgroundColor: 'lightgrey',
+  padding: '5px',
+  minHeight: '100px',
+  outline: 0,
+}
+
+class MentionsWithDropdown extends React.Component<{}, MentionsWithDropdownState> {
```
Comment (kuzhelov) [inline docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx:24]: this is just about visual representation - frankly, it was not absolutely evident to me that this grey box below is introduced as an _editable area_. 

image


I do understand that this was done to keep styles (and amount of code) minimal - and totally support that 💯 But maybe we could improve a bit in this aspect, by doing the following steps:

- change wording of description a bit to explicitly mention that area is editable
- use brighter styles that would differentiate area from regular rectangle

image


Please, feel free to decide on whether changes should be done, and which ones specifically. Let me just, for the sake of reference, share changes that I have tried on my side to achieve this appearance:

```js
const editorStyle: React.CSSProperties = {
  backgroundColor: '#eee',
  padding: '5px',
  minHeight: '100px',
  outline: 0,
  border: '1px dashed grey',
  borderRadius: '5px'
}
```

File: docs/src/routes.tsx [hunk 19]
```diff
@@ -62,9 +62,9 @@ const Router = () => (
           />,
           <DocsLayout
             exact
-            key="/prototype-async-dropdown-search"
-            path="/prototype-async-dropdown-search"
-            component={require('./prototypes/AsyncDropdownSearch/index').default}
+            key="/prototype-dropdowns"
```
Comment (kuzhelov) [inline docs/src/routes.tsx:65]: 👍

File: packages/react/src/components/Dropdown/Dropdown.tsx [hunk 20]
```diff
@@ -101,7 +101,10 @@ export interface DropdownProps extends UIComponentProps<DropdownProps, DropdownS
   items?: ShorthandCollection
 
   /**
-   * Function to be passed to create string from selected item, if it's a shorthand object. Used when dropdown also has a search function.
+   * Function that converts an item to string. Used when dropdown has the search boolean prop set to true.
```
Comment (kuzhelov) [inline packages/react/src/components/Dropdown/Dropdown.tsx:104]: 👍

File: docs/src/prototypes/Prototypes.tsx [hunk 21]
```diff
@@ -0,0 +1,29 @@
+import * as React from 'react'
+import { Box, Header, Segment } from '@stardust-ui/react'
+
+interface PrototypeSectionProps {
+  title?: string
+}
+
+interface ComponentPrototypeProps extends PrototypeSectionProps {
+  description?: string
```
Comment (layershifter) [inline docs/src/prototypes/Prototypes.tsx]:
```suggestion
  description?: React.ReactNode
```

File: docs/src/prototypes/Prototypes.tsx [hunk 22]
```diff
@@ -0,0 +1,29 @@
+import * as React from 'react'
+import { Box, Header, Segment } from '@stardust-ui/react'
+
+interface PrototypeSectionProps {
+  title?: string
```
Comment (layershifter) [inline docs/src/prototypes/Prototypes.tsx]:
```suggestion
  title?: React.ReactNode
```

Please allow to pass elements.

File: docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+        />
+        <PortalAtCursorPosition open={dropdownOpen}>
+          <Dropdown
+            defaultOpen={true}
+            inline
+            search
+            items={atMentionItems}
+            toggleIndicator={null}
+            searchInput={{
+              input: { autoFocus: true, size: searchQuery.length + 1 },
```
Comment (kuzhelov) [inline docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx:53]: seems that `size` prop won't work, as there is no `size` prop declared for `DropdownSearchInput` (or for the `Input`). Am I missing something? I think it would be safe for us just to omit this prop

File: docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+            onSearchQueryChange={this.handleSearchQueryChange}
+            noResultsMessage="We couldn't find any matches."
+          />
+        </PortalAtCursorPosition>
+      </>
+    )
+  }
+
+  private handleEditorKeyUp = (e: React.KeyboardEvent) => {
+    if (!this.state.dropdownOpen && keyboardKey.getCode(e) === keyboardKey.AtSign) {
```
Comment (kuzhelov) [inline docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx]: this condition will open dropdown even if `2` is typed - I would rather suggest to introduce simplest possible logic of checking for `@` symbol explicitly

File: docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+    dropdownOpen: false,
+    searchQuery: '',
+  }
+
+  private contendEditableRef = React.createRef<HTMLDivElement>()
+
+  state = this.initialState
+
+  render() {
+    const { dropdownOpen, searchQuery } = this.state
```
Comment (kuzhelov) [inline docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx]: cannot spot where `searchQuery` is used

File: packages/react/src/components/Dropdown/DropdownSearchInput.tsx (hunk 16 above)
Comment (kuzhelov) [reply packages/react/src/components/Dropdown/DropdownSearchInput.tsx:127]: share the same feeling as @layershifter expressed 😕

**Review by kuzhelov — COMMENTED** [review]
nice work 👍 just couple of things that I would like to be addressed before merge (apart from minor comments:
- sizing for `input` slot of dropdown input (could we remove this `size` altogether?): https://github.com/stardust-ui/react/pull/931/files#r259296667
- at-mentions dropdown is opened on `2` character typed: https://github.com/stardust-ui/react/pull/931/files#r259297611

File: docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx (hunk 17 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/AsyncDropdownSearch.tsx]: @kuzhelov `AsyncDropdownSearch.tsx` is not part of this PR;
 this file was just moved so that all dropdown related prototypes are in one place

File: docs/src/prototypes/Prototypes.tsx (hunk 21 above)
Comment (bmdalex) [reply docs/src/prototypes/Prototypes.tsx]: done, thx

File: docs/src/prototypes/Prototypes.tsx (hunk 22 above)
Comment (bmdalex) [reply docs/src/prototypes/Prototypes.tsx]: done, thx

File: docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx (hunk 23 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx:53]: `size` is an HTML attribute:

_unsigned long: Returns / Sets the element's size attribute, containing visual size of the control. This value is in pixels unless the value of type is text or password, in which case, it is an integer number of characters. Applies only when type is set to text, search, tel, url, email, or password; otherwise it is ignored.
(see: https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement)_

We need it to resize the input while typing; useful for mentioning people in the middle of text:
screen recording 2019-02-22 at 14 48 06

File: docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx (hunk 24 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx]: done

File: docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx (hunk 18 above)
Comment (bmdalex) [reply docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx:24]: I like it, applied the text change and style on content editable div, thanks!

**Comment by bmdalex** [comment]
thanks @kuzhelov , here are my responses:

> * sizing for `input` slot of dropdown input (could we remove this `size` altogether?): https://github.com/stardust-ui/react/pull/931/files#r259296667

see: https://github.com/stardust-ui/react/pull/931#discussion_r259348495

> * at-mentions dropdown is opened on `2` character typed: https://github.com/stardust-ui/react/pull/931/files#r259297611

fixed

File: docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx (hunk 23 above)
Comment (kuzhelov) [reply docs/src/prototypes/dropdowns/MentionsWithDropdown.tsx:53]: got it, thanks for explanation - actually, the fact that we are overlapping with DOM property's name seems a bit disturbing to me. But this is, definitely, a discussion aside from this PR.

**Review by kuzhelov — APPROVED** [review]
please, just introduce changelog entry before merging 👍
